### ISSUE TITLE
Use wp_enqueue_script to load web SDK

### ISSUE THREAD

**Issue by jasonpang** [OneSignal/OneSignal-WordPress-Plugin#84]
[Conversation](https://app.intercom.io/a/apps/344a89aeac3f033e4dec4370781543b948aece6d/inbox/conversation/8129243428)

We can use `wp_enqueue_script` to further increase WordPress site performance by placing our  SDK in the footer. The SDK is currently `async` loaded, but this still has the opportunity to block page rendering when it is executed.

**Comment by jasonpang** [comment]
Actually, I think our current integration is generally optimal. Focusing on these two points:

- Performance: Loading the JavaScript SDK asynchronously using `async` provides optimal performance for a majority of sites
- Compatibility: `wp_enqueue_scripts()` does not provide a way to load scripts asynchronously (not without a hack)

## Performance

It looks like we have the following choices when it comes to loading our script:

1. As a regular external `<script src="...">` tag
2. Using `defer`
3. Using `async`
4. Placing the script at the end of `<body>`
5. Loading the script after the page `load` event

### 1. `<script src="https://...OneSignalSDK.js">`
This is the worst.

- Performance: Worst
- Waits for CSS to be fully loaded? Yes <sup>[1]</sup>
- Stops page rendering to download the script? Yes
- Stops page parsing to execute the script? Yes
- Stops page rendering to execute the script? Yes
- Delays the page `load` event? Yes

### 2. `<script src="https://...OneSignalSDK.js" defer>`

Many sites suggest using `async` over `defer`. Looking at this diagram:

<img src="https://html.spec.whatwg.org/images/asyncdefer.svg"/>
<p align="center"><em>Source: https://html.spec.whatwg.org/multipage/scripting.html#the-script-element</em></p>

It seems like `defer` is preferable because:
- Document parsing is not interrupted before executing the script, whereas `async` executes the script whenever possible after downloading, even if it blocks document parsing

But a really interesting comment from Ilya's blog mentions:

> Initial load of most pages is mostly CPU idle: our measurement of top 1M Alexa sites shows 60% idle while waiting on the network. As a result, interleaving early script execution will help.
> 
> The reason the CPU is idle is not because of bandwidth contention, but because of roundtrip latency. Most pages peg the BW of their connection very infrequently and mostly when downloading a large asset far into the page load cycle (once the TCP CWND is high).
> 
> As a result, dispatching all of your requests as early as possible (JS and everything else), and allowing the browser to interleave the processing and execution is a win.

So it sounds like we actually _want_ scripts to execute as soon as they are downloaded since, for most sites, this conveniently occurs while waiting for other resources.

So, `defer` is _not preferable_ because:
- We want to execute scripts as they are finished downloading, to take advantage of idle CPU time while waiting for other network resources
- `defer` executes sequentially; if other scripts are marked `defer`, your script may be run a long time later. Our SDK is not dependent on any other resource, so this doesn't make sense.

Final tally:

- Waits for CSS to be fully loaded? Not sure, probably no (similar to `async)
- Stops page rendering to download the script? No
- Stops page parsing to execute the script? **No** (this was the only advantage from async). Executes before `DOMContentLoaded`.
- Stops page rendering to execute the script? Yes
- Delays the page `load` event? Yes
- Comments: Actually not preferable due to reasons above

### 3. `<script src="https://...OneSignalSDK.js" async>`

We're currently using this method. The resources below all recommend using `async` as well.

- Waits for CSS to be fully loaded? No
- Stops page rendering to download the script? No
- Stops page parsing to execute the script? Yes (advantageous for concurrency though)
- Stops page rendering to execute the script? Yes
- Delays the page `load` event? Yes (unavoidable without method #5)

Optimal solution, although it will delay the `load` event and stop to execute.

### 4. Script tag at the end of `<body>`

I'm not sure if the browser will peek ahead and prefetch this resource.

- Waits for CSS to be fully loaded? Yes (should be done by this point)
- Stops page rendering to download the script? Likely not (page is already rendered by this point maybe?)
- Stops page parsing to execute the script? Yes (advantageous for concurrency though)
- Stops page rendering to execute the script? Not sure. Does the page render sections as it is parsed? Or does it wait for everything to be parsed before rendering?
- Delays the page `load` event? Yes

Without prefetching, even though this is at the end of body, the browser waits to fetch this resource. This is bad! Doesn't block parsing (since there is no more at the end of `<body>`), which is good. With prefetching, isn't this like `async`? Not sure what the browser does, but `async` is recommended by many resources.

### 5. After page `load` event

- Waits for CSS to be fully loaded? No (CSS is already loaded)
- Stops page rendering to download the script? No (page is already rendered)
- Stops page parsing to execute the script? No (page is already parsed and rendered)
- Stops page rendering to execute the script? No (page is already rendered)
- Delays the page `load` event? **No** (occurs after)

See [this article](http://www.aaronpeters.nl/blog/why-loading-third-party-scripts-async-is-not-good-enough).

For users really trying to show page content ASAP and keeping down their `load` event time, this is the _only_ solution in this list that doesn't increase the page's load time. However, you have to wait for the `load` event. For some sites, this will take a long time and if we don't prompt for notifications until after, we could miss out on subscribers.

## Compatibility

- `wp_enqueue_script()` doesn't provide a way to mark a script `async`
- `script_loader_tag()` does, but is [introduced in WordPress 4.1.0 or higher])(https://developer.wordpress.org/reference/hooks/script_loader_tag)
- [A workaround exists](http://stackoverflow.com/a/20672324/555547)

I don't see any benefit to using `wp_enqueue_scripts()` if it takes a workaround to use `async`. Our current solution is working without any issues (except maybe one, see below), and using `wp_enqueue_scripts()` does not provide any _additional_ benefits.

## Things to Improve

If inline scripts wait for CSS to finish loading, maybe we should reference it in an external script tag with `async`? Although the CSS has to wait to finish loading anyways for any page content to be displayed.

## Reading Material

- https://www.igvita.com/2014/05/20/script-injected-async-scripts-considered-harmful/

  ^ This guy wrote the official Google Web Fundamentals article as well
- http://peter.sh/experiments/asynchronous-and-deferred-javascript-execution-explained/
- https://www.html5rocks.com/en/tutorials/speed/script-loading/ (from Jake Archibald)
- https://developers.google.com/web/fundamentals/performance/critical-rendering-path/adding-interactivity-with-javascript#parser-blocking-vs-asynchronous-javascript
- https://developers.google.com/speed/docs/insights/BlockingJS
- http://www.aaronpeters.nl/blog/why-loading-third-party-scripts-async-is-not-good-enough

[1]: https://www.igvita.com/2014/05/20/script-injected-async-scripts-considered-harmful
[2]: https://www.igvita.com/2014/05/20/script-injected-async-scripts-considered-harmful/#comment-1397785740

**Comment by grigs** [comment]
I forgot about the problem with `enqueue` not supporting async. If `script_loader_tag` supports it, that would be preferable. The current version of WP is 4.7.2 so something with 4.1.0 as a minimum seems fine.

I still think the fact that OneSignal is inserting itself into the head in a way that prevents the site owner from having control is undesirable. Take a look at this waterfall from our site:
http://www.webpagetest.org/result/170209_3R_5QP/2/details/#waterfall_view_step1

OneSignalSDK.js is the second resource the browser requests, but on my list of priorities, it would be 18th just above Google Analytics. In addition, you can see the CPU is pegged from before the SDK is requested so Ilya's comment about CPU idle during network requests doesn't apply here.

**Comment by amiremami** [comment]
I was searching on google how can I load onesignal in footer, and I found your post. Very amazing explanation. I hope you can improve it's loading.

Thanks :)

**Comment by mintplugins** [comment]
@jasonpang If you could take a second look here that would be great. It is possible to load using async and wp_enqueue_script by doing this:

```
function onesignal_enqueue_scripts() {
	wp_enqueue_script( 'onesignalsdk', 'https://cdn.onesignal.com/sdks/OneSignalSDK.js' );
}
add_action( 'wp_enqueue_scripts', 'onesignal_enqueue_scripts' );

function onesignal_modify_script_tag( $tag, $handle, $src ) {
  if ( 'onesignalsdk' == $handle ) {
    $tag = '<script src="https://cdn.onesignal.com/sdks/OneSignalSDK.js" async></script>';
  }
  return $tag;
}
add_filter( 'script_loader_tag', 'onesignal_modify_script_tag', 10, 3 );
```

This way, people who want to depend on the onesignalsdk can make sure it is loaded using the paramaters built into wp_enqueue_script.

**Comment by mintplugins** [comment]
Additionally, it would allow people to dequeue the script using wp_dequeue_script and re-enqueue it non asynchronously if they so choose. This would allow it to work the way WordPress recommends that plugins handle scripts and make it much more flexible for developers. Thanks for considering!
